kinto/plugins/quotas/listener.py

```diff
@@ -0,0 +1,212 @@
+                bucket_info['collection_count'] += 1
+                collection_info['storage_size'] += new_size
+            if resource_name == 'record':
+                bucket_info['record_count'] += 1
+                collection_info['record_count'] += 1
+                collection_info['storage_size'] += new_size
+        elif action == 'update':
+            bucket_info['storage_size'] -= old_size
+            bucket_info['storage_size'] += new_size
+            if resource_name in ('collection', 'record'):
+                collection_info['storage_size'] -= old_size
+                collection_info['storage_size'] += new_size
+        elif action == 'delete':
+            bucket_info['storage_size'] -= old_size
+            if resource_name == 'collection':
+                bucket_info['collection_count'] -= 1
+                # When we delete the collection all the records in it
+                # are deleted without notification.
+                collection_records, _ = storage.get_all('record',
+                                                        collection_uri)
+                for r in collection_records:
+                    old_record_size = record_size(r)
+                    bucket_info['record_count'] -= 1
+                    bucket_info['storage_size'] -= old_record_size
+                    collection_info['record_count'] -= 1
+                    collection_info['storage_size'] -= old_record_size
+                collection_info['storage_size'] -= old_size
+
+            if resource_name == 'record':
+                bucket_info['record_count'] -= 1
+                collection_info['record_count'] -= 1
+                collection_info['storage_size'] -= old_size
+
+    if bucket_max_bytes is not None:
+        if bucket_info['storage_size'] > bucket_max_bytes:
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=HTTPInsufficientStorage.explanation)
+
+    if bucket_max_items is not None:
+        if bucket_info['record_count'] > bucket_max_items:
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=HTTPInsufficientStorage.explanation)
+
+    if collection_max_bytes is not None:
+        if collection_info['storage_size'] > collection_max_bytes:
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=HTTPInsufficientStorage.explanation)
+
+    if collection_max_items is not None:
+        if collection_info['record_count'] > collection_max_items:
+            raise http_error(HTTPInsufficientStorage(),
+                             errno=ERRORS.FORBIDDEN.value,
+                             message=HTTPInsufficientStorage.explanation)
+
+    storage.update(parent_id=bucket_uri,
```

Comment:
> how do we protect the user from updating these records?

We prevent it because we don't have any API for the user to access it.
If we were to add this API we would need to make sure the user cannot modify it.